README.md
[![Crates.io](https://img.shields.io/crates/v/dcpu16-gui.svg)](https://crates.io/crates/dcpu16-gui)

# DCPU-16 GUI

DCPU-16 emulator GUI written in Rust and Piston.

## Installation

Make sure you have Rust/Cargo and SDL2. One of these lines might help with the latter:

    $ brew install sdl2                 # macOS
    $ apt-get install libsdl2-dev       # Ubuntu

Now install DCPU-16 and DCPU-16-GUI through Cargo:

    $ cargo install dcpu16
    $ cargo install dcpu16-gui

This will install a variety of binaries that all start with `dcpu16`, so try
typing that in and hit tab.

## Devices

Implemented devices:

* Monitor (LEM1802)
* Generic keyboard
* Floppy drive (M35FD)

They are attached with device indices 0 (monitor), 1 (keyboard), and 2
(floppy). The floppy drive is empty at start, but you have an all-zero floppy
disk that can be inserted/ejected by hitting F1.

## Example

Prints "Hello world!" in green to the screen:

    $ dcpu16-gui examples/hello.bin

To automatically attach a monitor to address `0x8000` (as in older programs),
run with `-m 0x8000`.

## Example 2

The next example is not assembled yet, so let's do that first:

    $ dcpu16-assembler examples/rainbow.dasm16 -o examples/rainbow.bin

Now we can run it:

    $ dcpu16-gui examples/rainbow.bin

## Tools

To assemble `dasm16` files into `bin` files, use [dcpu16](https://github.com/gustavla/dcpu16).


CHANGELOG.md
# Changelog

## Next release
Released: TBD
* Added Generic Keyboard (`DeviceKeyboardGeneric`)
  * The quotation and backtick keys are not triggering for some reason
* Added Floppy Drive (`DeviceFloppyM35FD`)
* Moved device implementations to separate files and directory (`src/devices`)
* Dependency update: dcpu16 0.3.0

## 0.2.0
Released: 2016-12-14
* LEM1802
  * Full support for all interrupts
  * Added border (can be configured through interrupts)
  * Supports blinking flag
  * Shows proper default screen when disconnected
  * Must specify custom address when pre-connecting monitor (e.g. `-m 0x8000`)
  * Ability to pre-configure font address via CLI (e.g. `-f 0x8180`)
* Dependency update: dcpu16 0.1.0

## 0.1.1
Released: 2016-12-06
* Dependency update: dcpu16 0.0.7

## 0.1.0
Released: 2016-12-06
* Updated to recent Piston version


src/floppy_m35fd.rs
use dcpu16::dcpu::{self, DCPU, Device};
use std::any::Any;
//use std::cmp;

const FLOPPY_SECTOR_SIZE: usize = 512;
const FLOPPY_NUM_SECTORS: usize = 1440;

// Read/write speed is 30700 words/second
// The DCPU-16 runs at 100 kHz, so a sector of size 512 words will take
// 512 / 30700 * 100000 = 1667 cycles to complete
const READ_WRITE_WAIT_CYCLES: usize = 1667;

// Cycles per track change (2.4 ms)
const READ_WRITE_TRACK_SEEK_CYCLES: usize = 240;

pub struct FloppyDisk {
    pub sectors: Vec<[u16; FLOPPY_SECTOR_SIZE]>,
    pub write_protected: bool,
}

impl FloppyDisk {
    pub fn new() -> FloppyDisk {
        FloppyDisk {
            sectors: Vec::new(),
            write_protected: false,
        }
    }
}

// This state is an implementation detail not related to the emulation
#[derive(Copy, Clone, Debug)]
enum FloppyInternalState {
    Idle,
    WaitToRead,
    WaitToWrite,
}

pub struct DeviceFloppyM35FD {
    state: u16,
    error: u16,
    pub interrupt_message: u16,
    pub disk: Option<FloppyDisk>,

    internal_state: FloppyInternalState,
    rw_sector: u16, // Also doubles as "last sector" before read/write
    rw_dcpu_address: u16,
    rw_wait_cycles: usize,

    interrupt_queued: bool,
}

const ERROR_NONE: u16       = 0x0000;
const ERROR_BUSY: u16       = 0x0001;
const ERROR_NO_MEDIA: u16   = 0x0002;
const ERROR_PROTECTED: u16  = 0x0003;
const ERROR_EJECT: u16      = 0x0004;
const ERROR_BAD_SECTOR: u16 = 0x0005;
const ERROR_BROKEN: u16     = 0xffff;

const STATE_NO_MEDIA: u16   = 0x0000;
const STATE_READY: u16      = 0x0001;
const STATE_READY_WP: u16   = 0x0002;
const STATE_BUSY: u16       = 0x0003;

impl DeviceFloppyM35FD {
    pub fn new() -> DeviceFloppyM35FD {
        DeviceFloppyM35FD {
            state: 0,
            error: 0,
            interrupt_message: 0,
            disk: None,

            internal_state: FloppyInternalState::Idle,
            rw_sector: 0, // Floppies generally start a cylinder 0
            rw_dcpu_address: 0x0,
            rw_wait_cycles: 0,

            interrupt_queued: false,
        }
    }

    pub fn state(&self) -> u16 {
        self.state
    }

    pub fn set_state(&mut self, state: u16) -> () {
        if self.state != state {
            self.interrupt_queued = true;
        }
        self.state = state;
    }

    pub fn error(&self) -> u16 {
        self.error
    }

    pub fn set_error(&mut self, error: u16) -> () {
        if self.error != error {
            self.interrupt_queued = true;
        }
        self.error = error;
    }

    /*
    pub fn has_disk(&self) -> bool {
        self.disk.is_some()
    }
    */

    pub fn insert(&mut self, disk: FloppyDisk) {
        // TODO: Check to see if floppy is already inserted?
        self.disk = Some(disk);
        self.set_state(STATE_READY);
    }

    pub fn eject(&mut self) -> Option<FloppyDisk> {
        self.set_state(STATE_NO_MEDIA);
        self.disk.take()
    }
}

impl Device for DeviceFloppyM35FD {
    fn info_hardware_id_upper(&self) -> u16 { 0x4fd5 }
    fn info_hardware_id_lower(&self) -> u16 { 0x24c5 }
    fn info_manufacturer_id_upper(&self) -> u16 { 0x1eb3 }
    fn info_manufacturer_id_lower(&self) -> u16 { 0x7e91 }
    fn info_version(&self) -> u16 { 0x000b }

    fn process_interrupt(&mut self, cpu: &mut DCPU) -> () {
        let a = cpu.reg[dcpu::REG_A];
        match a {
            0 => { // Poll device
                cpu.reg[dcpu::REG_B] = self.state() as u16;
                cpu.reg[dcpu::REG_C] = self.error() as u16;
                self.error = ERROR_NONE;
            },
            1 => { // Set interrupt
                let x = cpu.reg[dcpu::REG_X];
                self.interrupt_message = x;
            },
            2 => { // Read sector
                let x = cpu.reg[dcpu::REG_X];
                let y = cpu.reg[dcpu::REG_Y];

                let error = match self.state() {
                    STATE_NO_MEDIA => ERROR_NO_MEDIA,
                    STATE_BUSY => ERROR_BUSY,
                    STATE_READY | STATE_READY_WP => {
                        if (x as usize) < FLOPPY_NUM_SECTORS {
                            // Issue sector read, will be performed after delay in run()
                            self.internal_state = FloppyInternalState::WaitToRead;
                            let track1 = (self.rw_sector / 18) as isize;
                            self.rw_sector = x;
                            let track2 = (self.rw_sector / 18) as isize;
                            self.rw_dcpu_address = y;
                            self.rw_wait_cycles = READ_WRITE_WAIT_CYCLES;
                            let track_diff = (track1 - track2).abs() as usize;
                            // Code if we want to switch to two-sided floppy with 160 tracks
                            // We assume a two-sided floppy (without independent seeks), so seek
                            // distance is zero between tracks 0 and 79, and similarly 5 and 75.
                            //let track_diff = cmp::min((track1 - track2).abs(),
                            //                          (159 - track1 - track2).abs()) as usize;
                            self.rw_wait_cycles += track_diff * READ_WRITE_TRACK_SEEK_CYCLES;
                            self.set_state(STATE_BUSY);
                            ERROR_NONE
                        } else {
                            ERROR_BAD_SECTOR
                        }
                    },
                    _ => ERROR_BROKEN,
                };
                self.set_error(error);
            },
            3 => { // Write sector
                let x = cpu.reg[dcpu::REG_X];
                let y = cpu.reg[dcpu::REG_Y];

                let error = match self.state() {
                    STATE_NO_MEDIA => ERROR_NO_MEDIA,
                    STATE_READY_WP => ERROR_PROTECTED,
                    STATE_BUSY => ERROR_BUSY,
                    STATE_READY => {
                        if (x as usize) < FLOPPY_NUM_SECTORS {
                            // Issue sector write, will be performed after delay in run()
                            self.internal_state = FloppyInternalState::WaitToWrite;
                            self.rw_sector = x;
                            self.rw_dcpu_address = y;
                            self.rw_wait_cycles = READ_WRITE_WAIT_CYCLES;
                            self.set_state(STATE_BUSY);
                            ERROR_NONE
                        } else {
                            ERROR_BAD_SECTOR
                        }
                    },
                    _ => ERROR_BROKEN,
                };
                self.set_error(error);
            }
            _ => {
                // Do nothing
            },
        }
    }

    fn run(&mut self, cpu: &mut DCPU, cycle: usize) -> () {
        match self.internal_state {
            FloppyInternalState::WaitToRead => {
                if self.rw_wait_cycles > cycle {
                    self.rw_wait_cycles -= cycle;
                } else {
                    self.internal_state = FloppyInternalState::Idle;
                    self.rw_wait_cycles = 0;
                    let mut state = self.state();
                    let error;
                    match self.disk {
                        Some(ref floppy_disk) => {
                            match floppy_disk.sectors.get(self.rw_sector as usize) {
                                Some(s) => {
                                    for i in 0..FLOPPY_SECTOR_SIZE {
                                        cpu.mem[self.rw_dcpu_address.wrapping_add(i as u16) as usize] = s[i];
                                    }
                                },
                                None => {
                                    for i in 0..FLOPPY_SECTOR_SIZE {
                                        cpu.mem[self.rw_dcpu_address.wrapping_add(i as u16) as usize] = 0;
                                    }
                                },
                            }
                            state = match floppy_disk.write_protected {
                                true => STATE_READY_WP,
                                false => STATE_READY,
                            };
                            error = ERROR_NONE;
                        },
                        None => {
                            error = ERROR_EJECT;
                        },
                    }
                    self.set_state(state);
                    self.set_error(error);
                }
            },
            FloppyInternalState::WaitToWrite => {
                if self.rw_wait_cycles > cycle {
                    self.rw_wait_cycles -= cycle;
                } else {
                    self.internal_state = FloppyInternalState::Idle;
                    self.rw_wait_cycles = 0;
                    let mut state = self.state();
                    let error;
                    match self.disk {
                        Some(ref mut floppy_disk) => {
                            // Make sure sector exists
                            while floppy_disk.sectors.len() <= (self.rw_sector as usize) {
                                floppy_disk.sectors.push([0; FLOPPY_SECTOR_SIZE]);
                            }
                            match floppy_disk.sectors.get_mut(self.rw_sector as usize) {
                                Some(ref mut s) => {
                                    for i in 0..FLOPPY_SECTOR_SIZE {
                                        s[i as usize] = cpu.mem[self.rw_dcpu_address.wrapping_add(i as u16) as usize];
                                    }
                                },
                                None => {
                                    unreachable!();
                                },
                            }

                            // TODO: Potentially do a sector clean-up if trailing sectors
                            // are all-zero

                            state = match floppy_disk.write_protected {
                                true => STATE_READY_WP,
                                false => STATE_READY,
                            };
                            error = ERROR_NONE;
                        },
                        None => {
                            error = ERROR_EJECT;
                        },
                    }
                    self.set_state(state);
                    self.set_error(error);
                }
            },
            FloppyInternalState::Idle => {},
        }

        if self.interrupt_queued {
            if self.interrupt_message != 0 {
                cpu.interrupt(self.interrupt_message);
            }
            self.interrupt_queued = false;
        }
    }

    fn as_any(&self) -> &Any {
        self
    }

    fn as_any_mut(&mut self) -> &mut Any {
        self
    }
}


src/devices/monitor_lem1802.rs
use dcpu16::dcpu::{self, DCPU, Device};
use std::any::Any;

pub struct DeviceMonitorLEM1802 {
    pub connected: bool,
    pub ram_location: u16,
    pub font_location: Option<u16>,
    pub palette_location: Option<u16>,
    pub border_color_index: u16,
}

pub const MONITOR_WIDTH: u32 = COLS as u32 * FONT_WIDTH as u32;
pub const MONITOR_HEIGHT: u32 = ROWS as u32 * FONT_HEIGHT as u32;
pub const ROWS: usize = 12;
pub const COLS: usize = 32;
pub const FONT_WIDTH: usize = 4;
pub const FONT_HEIGHT: usize = 8;
pub const SCALE: u32 = 5;
pub const BORDER: u32 = 5*SCALE;

const DEFAULT_FONT: &'static [u16] = &[
    0xb79e, 0x388e, 0x722c, 0x75f4, 0x19bb, 0x7f8f, 0x85f9, 0xb158, 0x242e, 0x2400, 0x082a, 0x0800,
    0x0008, 0x0000, 0x0808, 0x0808, 0x00ff, 0x0000, 0x00f8, 0x0808, 0x08f8, 0x0000, 0x080f, 0x0000,
    0x000f, 0x0808, 0x00ff, 0x0808, 0x08f8, 0x0808, 0x08ff, 0x0000, 0x080f, 0x0808, 0x08ff, 0x0808,
    0x6633, 0x99cc, 0x9933, 0x66cc, 0xfef8, 0xe080, 0x7f1f, 0x0701, 0x0107, 0x1f7f, 0x80e0, 0xf8fe,
    0x5500, 0xaa00, 0x55aa, 0x55aa, 0xffaa, 0xff55, 0x0f0f, 0x0f0f, 0xf0f0, 0xf0f0, 0x0000, 0xffff,
    0xffff, 0x0000, 0xffff, 0xffff, 0x0000, 0x0000, 0x005f, 0x0000, 0x0300, 0x0300, 0x3e14, 0x3e00,
    0x266b, 0x3200, 0x611c, 0x4300, 0x3629, 0x7650, 0x0002, 0x0100, 0x1c22, 0x4100, 0x4122, 0x1c00,
    0x1408, 0x1400, 0x081c, 0x0800, 0x4020, 0x0000, 0x0808, 0x0800, 0x0040, 0x0000, 0x601c, 0x0300,
    0x3e49, 0x3e00, 0x427f, 0x4000, 0x6259, 0x4600, 0x2249, 0x3600, 0x0f08, 0x7f00, 0x2745, 0x3900,
    0x3e49, 0x3200, 0x6119, 0x0700, 0x3649, 0x3600, 0x2649, 0x3e00, 0x0024, 0x0000, 0x4024, 0x0000,
    0x0814, 0x2241, 0x1414, 0x1400, 0x4122, 0x1408, 0x0259, 0x0600, 0x3e59, 0x5e00, 0x7e09, 0x7e00,
    0x7f49, 0x3600, 0x3e41, 0x2200, 0x7f41, 0x3e00, 0x7f49, 0x4100, 0x7f09, 0x0100, 0x3e41, 0x7a00,
    0x7f08, 0x7f00, 0x417f, 0x4100, 0x2040, 0x3f00, 0x7f08, 0x7700, 0x7f40, 0x4000, 0x7f06, 0x7f00,
    0x7f01, 0x7e00, 0x3e41, 0x3e00, 0x7f09, 0x0600, 0x3e41, 0xbe00, 0x7f09, 0x7600, 0x2649, 0x3200,
    0x017f, 0x0100, 0x3f40, 0x3f00, 0x1f60, 0x1f00, 0x7f30, 0x7f00, 0x7708, 0x7700, 0x0778, 0x0700,
    0x7149, 0x4700, 0x007f, 0x4100, 0x031c, 0x6000, 0x0041, 0x7f00, 0x0201, 0x0200, 0x8080, 0x8000,
    0x0001, 0x0200, 0x2454, 0x7800, 0x7f44, 0x3800, 0x3844, 0x2800, 0x3844, 0x7f00, 0x3854, 0x5800,
    0x087e, 0x0900, 0x4854, 0x3c00, 0x7f04, 0x7800, 0x447d, 0x4000, 0x2040, 0x3d00, 0x7f10, 0x6c00,
    0x417f, 0x4000, 0x7c18, 0x7c00, 0x7c04, 0x7800, 0x3844, 0x3800, 0x7c14, 0x0800, 0x0814, 0x7c00,
    0x7c04, 0x0800, 0x4854, 0x2400, 0x043e, 0x4400, 0x3c40, 0x7c00, 0x1c60, 0x1c00, 0x7c30, 0x7c00,
    0x6c10, 0x6c00, 0x4c50, 0x3c00, 0x6454, 0x4c00, 0x0836, 0x4100, 0x0077, 0x0000, 0x4136, 0x0800,
    0x0201, 0x0201, 0x0205, 0x0200
];

const DEFAULT_PALETTE: &'static [u16] = &[
    0x0000, 0x000a, 0x00a0, 0x00aa, 0x0a00, 0x0a0a, 0x0a50, 0x0aaa,
    0x0555, 0x055f, 0x05f5, 0x05ff, 0x0f55, 0x0f5f, 0x0ff5, 0x0fff,
];

// Logo pixels, describes row-major values where pixel is yellow
const LOGO_PIXELS: &'static [usize] = &[
    3379, 3389, 3507, 3517, 3636, 3645, 3764, 3773, 3893, 3901, 4014,
    4021, 4029, 4142, 4150, 4157, 4270, 4271, 4278, 4285, 4398, 4399,
    4407, 4413, 4422, 4423, 4424, 4425, 4426, 4427, 4428, 4429, 4430,
    4431, 4432, 4433, 4526, 4528, 4535, 4541, 4654, 4656, 4664, 4669,
    4782, 4785, 4792, 4797, 4910, 4913, 4921, 4925, 5038, 5042, 5049,
    5053, 5166, 5170, 5178, 5181, 5294, 5299, 5306, 5309, 5422, 5427,
    5435, 5437, 5550, 5556, 5563, 5565, 5678, 5684, 5692, 5693, 5702,
    5703, 5704, 5705, 5706, 5707, 5708, 5709, 5710, 5711, 5712, 5713,
    5806, 5813, 5820, 5821, 5934, 5941, 5949, 6062, 6070, 6077, 6190,
    6198, 6318, 6327, 6446, 6455, 6574, 6584, 6702, 6712, 7462, 7463,
    7466, 7468, 7471, 7472, 7477, 7478, 7479, 7481, 7485, 7486, 7487,
    7489, 7491, 7493, 7494, 7495, 7497, 7498, 7499, 7501, 7504, 7505,
    7507, 7509, 7512, 7513, 7590, 7592, 7594, 7595, 7596, 7598, 7599,
    7600, 7605, 7606, 7609, 7613, 7614, 7617, 7618, 7622, 7625, 7626,
    7629, 7632, 7635, 7636, 7639, 7640, 7641, 7718, 7720, 7723, 7726,
    7728, 7733, 7734, 7735, 7737, 7738, 7739, 7741, 7742, 7743, 7745,
    7747, 7750, 7753, 7755, 7757, 7759, 7760, 7763, 7765, 7767, 7769,
];

impl DeviceMonitorLEM1802 {
    pub fn new() -> DeviceMonitorLEM1802 {
        DeviceMonitorLEM1802 {
            connected: false,
            ram_location: 0,
            font_location: None,
            palette_location: None,
            border_color_index: 0,
        }
    }

    pub fn with_pre_connect(self, location: u16) -> DeviceMonitorLEM1802 {
        let mut new_self = self;
        new_self.connected = true;
        new_self.ram_location = location;
        new_self
    }

    pub fn with_font_location(self, location: u16) -> DeviceMonitorLEM1802 {
        let mut new_self = self;
        new_self.font_location = Some(location);
        new_self
    }

    /*
    pub fn with_palette_location(self, location: u16) -> DeviceMonitorLEM1802 {
        let mut new_self = self;
        new_self.palette_location = Some(location);
        new_self
    }

    pub fn with_border_color_index(self, border_color_index: u16) -> DeviceMonitorLEM1802 {
        let mut new_self = self;
        new_self.border_color_index = border_color_index & 0xf;
        new_self
    }
    */

    pub fn get_font_character(&self, cpu: &DCPU, c: usize) -> u32 {
        match self.font_location {
            Some(loc) => {
                ((cpu.mem[(loc as usize + c * 2) % dcpu::MEMORY_SIZE] as u32) << 16) +
                 (cpu.mem[(loc as usize + c * 2 + 1) % dcpu::MEMORY_SIZE] as u32)
            },
            None => {
                ((DEFAULT_FONT[c * 2] as u32) << 16) +
                 (DEFAULT_FONT[c * 2 + 1] as u32)
            }
        }
    }

    pub fn get_color(&self, cpu: &DCPU, color_index: u16) -> u16 {
        match self.palette_location {
            Some(loc) => {
                cpu.mem[loc.wrapping_add(color_index) as usize]
            },
            None => {
                DEFAULT_PALETTE[color_index as usize]
            },
        }
    }

    pub fn get_border_color_rgb(&self, cpu: &DCPU) -> (u8, u8, u8) {
        if self.connected {
            let color = self.get_color(cpu, self.border_color_index);
            (
                ((((color >> 8) & 0xf) << 4) | ((color >> 8) & 0xf)) as u8,
                ((((color >> 4) & 0xf) << 4) | ((color >> 4) & 0xf)) as u8,
                ((((color     ) & 0xf) << 4) | ((color     ) & 0xf)) as u8,
            )
        } else {
            (0, 0, 170)
        }
    }

    pub fn data(&self, cpu: &DCPU, blinkout: bool) -> Vec<u8> {
        let mut v: Vec<u8> = vec![0; (MONITOR_WIDTH * MONITOR_HEIGHT * 3) as usize];
        {
            let mut slice = &mut v[..];
            if self.connected {
                for i in 0..ROWS {
                    for j in 0..COLS {
                        let mem = cpu.mem[(self.ram_location as usize + i * COLS + j) % dcpu::MEMORY_SIZE];
                        let c = (mem & 0x7f) as usize;

                        let blink = ((mem >> 7) & 1) == 1;
                        let bg_color_index = ((mem >> 8) & 0xf) as u16;
                        let fg_color_index = ((mem >> 12) & 0xf) as u16;

                        let bg_color = self.get_color(cpu, bg_color_index);
                        let fg_color = self.get_color(cpu, fg_color_index);

                        let b0 = if !blinkout || !blink {
                            self.get_font_character(cpu, c)
                        } else {
                            0
                        };

                        for x in 0..FONT_WIDTH {
                            for y in 0..FONT_HEIGHT {
                                let p = (b0 >> ((FONT_WIDTH-1-x) * FONT_HEIGHT + y)) & 1;
                                let color = match p {
                                    1 => fg_color,
                                    _ => bg_color,
                                };

                                let index = i * COLS * FONT_WIDTH * FONT_HEIGHT + y * COLS * FONT_WIDTH + j * FONT_WIDTH + x;
                                slice[index * 3    ] = ((((color >> 8) & 0xf) << 4) | ((color >> 8) & 0xf)) as u8;
                                slice[index * 3 + 1] = ((((color >> 4) & 0xf) << 4) | ((color >> 4) & 0xf)) as u8;
                                slice[index * 3 + 2] = ((((color     ) & 0xf) << 4) | ((color     ) & 0xf)) as u8;
                            }
                        }
                    }
                }
            } else {
                // Clear screen to blue
                for i in 0..((MONITOR_WIDTH * MONITOR_HEIGHT) as usize) {
                    slice[i * 3    ] = 0;
                    slice[i * 3 + 1] = 0;
                    slice[i * 3 + 2] = 170;
                }

                // Print logo and "Nya Elektriska" in yellow
                for p in LOGO_PIXELS {
                    slice[p * 3    ] = 255;
                    slice[p * 3 + 1] = 255;
                    slice[p * 3 + 2] = 0;
                }
            }
        }
        v
    }
}


impl Device for DeviceMonitorLEM1802 {
    fn info_hardware_id_upper(&self) -> u16 { 0x7349 }
    fn info_hardware_id_lower(&self) -> u16 { 0xf615 }
    fn info_manufacturer_id_upper(&self) -> u16 { 0x1c6c }
    fn info_manufacturer_id_lower(&self) -> u16 { 0x8b36 }
    fn info_version(&self) -> u16 { 0x1802 }

    fn process_interrupt(&mut self, cpu: &mut DCPU) -> () {
        let a = cpu.reg[0];
        let b = cpu.reg[1];
        match a {
            0 => { /* MEM_MAP_SCREEN */
                if b > 0 {
                    self.ram_location = b;
                    self.connected = true;
                } else {
                    self.connected = false;
                }
                // TODO: Disable screen for a 1 second
            },
            1 => { /* MEM_MAP_FONT */
                if b > 0 {
                    self.font_location = Some(b);
                } else {
                    self.font_location = None;
                }
            },
            2 => { /* MEM_MAP_PALETTE */
                if b > 0 {
                    self.palette_location = Some(b);
                } else {
                    self.palette_location = None;
                }
            },
            3 => { /* SET_BORDER_COLOR */
                self.border_color_index = b & 0xf;
            },
            4 => { /* MEM_DUMP_FONT */
                for i in 0..DEFAULT_FONT.len() {
                    cpu.mem[b.wrapping_add(i as u16) as usize] = DEFAULT_FONT[i];
                }
                cpu.halt(256);
            }
            5 => { /* MEM_DUMP_PALETTE */
                for i in 0..DEFAULT_PALETTE.len() {
                    cpu.mem[b.wrapping_add(i as u16) as usize] = DEFAULT_PALETTE[i];
                }
                cpu.halt(16);
            }
            _ => {}
        }
    }

    fn run(&mut self, _: &mut DCPU, _: usize) -> () {}

    fn as_any(&self) -> &Any {
        self
    }

    fn as_any_mut(&mut self) -> &mut Any {
        self
    }
}




src/main.rs
extern crate dcpu16;
extern crate piston;
extern crate graphics;
extern crate opengl_graphics;
extern crate gl;
extern crate libc;
extern crate getopts;
extern crate piston_window;
extern crate image;
extern crate event_loop;
extern crate time;

mod cli;
mod devices;

use std::path::Path;
use std::env;
use dcpu16::dcpu::DCPU;
use getopts::Options;
use event_loop::{Events, EventLoop};
use graphics::Transformed;
use piston::input::{RenderEvent, Button, PressEvent, ReleaseEvent};
use piston::input::keyboard::Key;

use devices::monitor_lem1802::{DeviceMonitorLEM1802, MONITOR_WIDTH, MONITOR_HEIGHT, SCALE, BORDER};
use devices::keyboard_generic::DeviceKeyboardGeneric;
use devices::floppy_m35fd::{DeviceFloppyM35FD, FloppyDisk};

fn main() {
    let mut opts = Options::new();
    let args: Vec<String> = env::args().collect();
    let program = args[0].clone();

    opts.optopt("m", "monitor", "Pre-connect monitor", "ADDRESS");
    opts.optopt("f", "font", "Pre-configure font", "ADDRESS");
    opts.optflag("v", "version", "print version");
    opts.optflag("h", "help", "print this help menu");
    let matches = match opts.parse(&args[1..]) {
        Ok(m) => m,
        Err(why) => {
            println!("{}", why);
            return;
        },
    };

    if matches.opt_present("h") {
        cli::print_usage(&program, "FILE", opts,
                         &["examples/rainbow.bin", "-m 0x8000 -f 0x8180 old-program.bin"]);
        return;
    }

    if matches.opt_present("v") {
        cli::print_version(&program);
        return;
    }

    if matches.free.len() != 1 {
        println!("Please input file");
        return;
    }
    let ref filename = matches.free[0];

    let path = Path::new(filename);
    let window_size = piston_window::Size {width: MONITOR_WIDTH * SCALE + BORDER * 2 as u32, height: (MONITOR_HEIGHT as u32 * SCALE + BORDER * 2)};

    let mut window: piston_window::PistonWindow = piston_window::WindowSettings::new(
            "DCPU16 Monitor",
            window_size
        )
        .exit_on_esc(true)
        .opengl(piston_window::OpenGL::V3_2)
        .build()
        .unwrap();

    let mut cpu = DCPU::new();
    if let Err(why) = cpu.load_from_binary_file(&path) {
        println!("Could load file {}: {}", path.display(), why);
        return;
    }

    let mut monitor = DeviceMonitorLEM1802::new();

    monitor = match matches.opt_str("monitor") {
        Some(s) => {
            if s.len() >= 2 && s.chars().nth(0).unwrap() == '0' && (s.chars().nth(1).unwrap() == 'X' || s.chars().nth(1).unwrap() == 'x') {
                match isize::from_str_radix(&s[2..], 16) {
                    Ok(loc) => monitor.with_pre_connect(loc as u16),
                    Err(e) => {
                        println!("Could not parse argument for -m/--monitor: {}", e);
                        return;
                    },
                }
            } else {
                println!("Could not parse argument for -m/--monitor: must be hexadecimal with leading 0x");
                return;
            }
        },
        None => monitor,
    };
    // TODO Remove this hideous copy-paste from above
    monitor = match matches.opt_str("font") {
        Some(s) => {
            if s.len() >= 2 && s.chars().nth(0).unwrap() == '0' && (s.chars().nth(1).unwrap() == 'X' || s.chars().nth(1).unwrap() == 'x') {
                match isize::from_str_radix(&s[2..], 16) {
                    Ok(loc) => monitor.with_font_location(loc as u16),
                    Err(e) => {
                        println!("Could not parse argument for -f/--font: {}", e);
                        return;
                    },
                }
            } else {
                println!("Could not parse argument for -f/--font: must be hexadecimal with leading 0x");
                return;
            }
        },
        None => monitor,
    };

    cpu.add_device(Box::new(monitor));

    let keyboard = DeviceKeyboardGeneric::new();
    cpu.add_device(Box::new(keyboard));

    let floppy_drive = DeviceFloppyM35FD::new();
    let mut disk = Some(FloppyDisk::new());

    cpu.add_device(Box::new(floppy_drive));

    // TODO: Make devices configurable through CLI instead

    let mut img = image::ImageBuffer::new(MONITOR_WIDTH as u32, MONITOR_HEIGHT as u32);

    let mut text_sett = piston_window::TextureSettings::new();
    text_sett.set_filter(piston_window::texture::Filter::Nearest);
    let mut texture = piston_window::Texture::from_image(&mut window.factory, &mut img, &text_sett).unwrap();

    window.set_bench_mode(true);
    let mut events = window.events().max_fps(30);
    while let Some(e) = events.next(&mut window) {
        let time = time::get_time();
        let blinkout = time.nsec > 500_000_000;

        if let Some(Button::Keyboard(key)) = e.press_args() {

            //if let Some(devices) = Rc::get_mut(&mut cpu.devices) {
            let devices = cpu.devices.clone();
            {
                let mut dev = devices.get(1).unwrap().borrow_mut();
                if let Some(mut keyboard) = dev.as_any_mut().downcast_mut::<DeviceKeyboardGeneric>() {
                    let v = devices::keyboard_generic::piston_key_to_code(key);
                    if v > 0 {
                        keyboard.register_press(&mut cpu, v);
                    }
                }
            }

            // Toggle insert/eject floppy
            if key == Key::F1 {
                let devices = cpu.devices.clone();
                {
                    let mut dev = devices.get(2).unwrap().borrow_mut();
                    if let Some(mut drive) = dev.as_any_mut().downcast_mut::<DeviceFloppyM35FD>() {
                        if disk.is_some() {
                            let disk0 = disk.take();
                            drive.insert(disk0.unwrap());
                            println!("Inserted floppy disk");
                        } else {
                            disk = drive.eject();
                            println!("Ejected floppy disk");
                        }
                    }
                }
            }
        }
        if let Some(Button::Keyboard(key)) = e.release_args() {
            let devices = cpu.devices.clone();
            {
                let mut dev = devices.get(1).unwrap().borrow_mut();
                if let Some(mut keyboard) = dev.as_any_mut().downcast_mut::<DeviceKeyboardGeneric>() {
                    let v = devices::keyboard_generic::piston_key_to_code(key);
                    if v > 0 {
                        keyboard.register_release(&mut cpu, v);
                    }
                }
            }
        }
        if let Some(_) = e.render_args() {
            // With FPS at 30, this will make the DCPU-16 run at 100 kHz
            cpu.run(3333);
            let dev = cpu.devices.get(0).unwrap().borrow();
            if let Some(monitor) = dev.as_any().downcast_ref::<DeviceMonitorLEM1802>() {
                let v = &monitor.data(&cpu, blinkout)[..];
                for k in 0..(v.len() / 3) {
                    let color = image::Rgba([v[3 * k], v[3 * k + 1], v[3 * k + 2], 255]);
                    img.put_pixel((k as u32 % MONITOR_WIDTH), k as u32 / MONITOR_WIDTH as u32, color);
                }

                match texture.update(&mut window.encoder, &img) {
                    Ok(_) => {},
                    Err(_) => {
                        println!("Error");
                    }
                }
            }
        }
        window.draw_2d(&e, |c, g| {
            let dev = cpu.devices.get(0).unwrap().borrow();
            if let Some(monitor) = dev.as_any().downcast_ref::<DeviceMonitorLEM1802>() {
                let (col_r, col_g, col_b) = monitor.get_border_color_rgb(&cpu);
                piston_window::clear([col_r as f32 / 255.0, col_g as f32 / 255.0, col_b as f32 / 255.0, 1.0], g);

                piston_window::image(&texture, c.transform
                    .trans(BORDER as f64, BORDER as f64).zoom(SCALE as f64), g);
            }
        });
    }
}
